keras/engine/training.py

```diff
@@ -962,10 +962,21 @@ def _make_predict_function(self):
                                                name='predict_function',
                                                **kwargs)
 
+    def _check_num_samples(self, ins, batch_size=None, steps=None, steps_name='steps'):
+        if steps is not None:
+            num_samples = steps
```

Comment:
We need to separate two use cases cleanly:
- training for a specific number of steps (if the data has no length), like in `fit_generator`.
- training for a specific number of samples (if the data has a length), like in current `fit`.

We can't unify both because no reliable conversion exists between the two.